File: LeetCode/BinaryTreeInorderTraversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ArrayList<Integer> inorderTraversal(TreeNode root) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        ArrayList<Integer> ans = new ArrayList<Integer>();
        if(root == null) return ans;
        HashSet<TreeNode> visited = new HashSet<TreeNode>();
        Stack<TreeNode> stack = new Stack();
        stack.push(root);
        while(!stack.empty()) {
            TreeNode node = stack.pop();
            if(visited.contains(node))
                ans.add(node.val);
            else {
                visited.add(node);
                if(node.right != null)
                    stack.push(node.right);
                stack.push(node);
                if(node.left != null)
                    stack.push(node.left);
            }
        }
        return ans;
    }
}

File: LeetCode/BinaryTreeInorderTraversal.cc
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *p;
    stack<TreeNode*> path;
    void goLeft() {
        path.push(p);
        p = p->left;
    }
    void goRight() {
        path.push(p);
        p = p->right;
    }
    void goUp() {
        p = path.top();
        path.pop();
    }

    vector<int> inorderTraversal(TreeNode *root) {
        vector<int> ans;
        p = root;
        while(!path.empty()) path.pop();
        if(!p) {
            return ans;
        }
        while(p->left) {
            goLeft();
        }

        do {
            ans.push_back(p->val);
        } while(next());
        return ans;
    }
    
    bool next() {
        if(p->right) {
            goRight();
            while(p->left) {
                goLeft();
            }
            return true;
        } else {
            while(!path.empty() && p == path.top()->right) {
                goUp();
            }
            if(path.empty()) {
                return false;
            } else {
                goUp();
                return true;
            }
        }
    }
};